# **Tarea 1 - CC6205 Natural Language Processing 📚**

**Integrantes:** Arturo Lazcano

**Fecha límite de entrega 📆:** Martes 4 de Abril.

**Tiempo estimado de dedicación:** 




` ` 


Bienvenid@s a la primera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las primeras semanas de clases, enfocado principalmente en ***Information Retrieval (IR)*** y ***Vector Space Models***. Si aún no han visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una parte teórica que busca evaluar conceptos vistos en clases. Seguido por una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP. 

` ` 




**Instrucciones:**
- La tarea se realiza en grupos de **máximo** 2 personas. Puede ser invidivual pero no es recomendable.
- La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
- El formato de entrega es este mismo Jupyter Notebook.
- Al momento de la revisión tu código será ejecutado. Por favor verifica que tu entrega no tenga errores de compilación. 
- Está **PROHIBIDO** usar cualquier librería que implemente los algoritmos pedidos (Spacy, scikit, etc). Sólo se podrán utilizar las librerías importadas al inicio de la sección de práctica.
- En el horario de auxiliar pueden realizar consultas acerca de la tarea a través del canal de Discord del curso. 



Ahora sí, empecemos! 😄🚀

` ` 

**Referencias:**


Slides:
    
- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Vector Space Model / Information Retrieval](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)    

Videos: 

- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte I](https://www.youtube.com/watch?v=HEKTNOttGvU)
- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte II](https://www.youtube.com/watch?v=P8cwnI-f-Kg)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 1](https://youtu.be/FXIVClF370w)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 2](https://youtu.be/f8nG1EMmPZk)

` ` 



## **1 - Preguntas teóricas 📕 (2 puntos).** ##









Las siguientes celdas contienen preguntas acerca del contenido visto en clases y en el material del curso.  Contestar cada pregunta en su celda correspondiente y **no extenderse más de 100 palabras** . 🙏
` ` 

**Pregunta 1 (0.2 puntos): ¿Por qué el análisis del lenguaje humano es una tarea compleja? Mencione dos razones según lo visto en clases. Debe citar la slide donde cree que está la respuesta.**




` ` 
El lenguaje es difìcil de analizar pues su significado es puede ser altamente distinto con tan solo una palabra o incluso un símbolo distinto, por ejemplo, una coma "," puesta en otro lugar. Interpertar el cambio de significado que puede tener una oración es una tarea interpretativa, y así, compleja y desafiante.  
Por otro lado, el lenguaje evoluciona con el tiempo, y esta evolución no sigue reglas concretas, solamente se adecúa al entorno y la epoca en la cual se está viviendo.  
Estas ideas están presentes en las slides "Linguistic levels of description" y "Challenges in NLP" ya que puede estudiarse bien un dataset, sin embargo, puede no ser generalizable a otros.
` ` 

**Pregunta 2 (0.4 puntos): ¿Cuáles son las diferencias entre usar Deep learning y Machine Learning clásico (empirismo) para un problema de NLP? Ejemplifique con alguna task.** Puede utilizar ChatGPT (debe indicarlo) para generar la respuesta y luego debe indicar si la respuesta entregada por ChatGPT es correcta o no. Mencione por qué según lo visto en clases.


` ` 
El machine learning clásico para NLP no era tan robusto como lo es deeplearning ahora mismo. El enfoque y mayor esfuerzo antiguamente era buscar las buenas reglas y features que pudiesen mejor resolver el problema. Por otro lado, el uso de deeplearning cambia esto por algoritmos de redes neuronaes que obtienen sus propios patrones y no vienen alimentados de estos "buenos features" por parte de la persona que lo ejecuta.  
Un ejemplo visto en clases es el de representar emociones vía tweets.
` ` 

 **Pregunta 3 (0.4 puntos) Según las primeras clases, ¿Qué método clásico nos permite rankear las similitudes existentes entre documentos?, ¿Cómo son las representaciones que genera y problemas que podrían experimentar estas soluciones simples?** 

` ` 
El modelo usado visto en la slide 2 del curso es el Vector Space Model. Se ven representados mediante vectores de clases de *tokens* normalizados. Así, se genera un espacio donde cada término es una dimensión de tal vector.  
Un problema de este modelo es que pierde estructura del lenguaje y que cada vector puede tener una gran dimensionalidad haciendo difícil la tarea computacionalmemte.
` ` 

**Pregunta 4 (0.4 puntos) Usted se encuentra realizando un modelo de clasificación de sentimientos con texto, su jefe le señala que debe eliminar las palabras mas comunes para obtener una mejor clasificación. ¿Qué palabras le señala que elimine su jefe?, ¿es acaso esto una buena idea?.** Puede utilizar ChatGPT (debe indicarlo) para generar la respuesta y luego debe indicar si la respuesta entregada por ChatGPT es correcta o no. Mencione por qué según lo visto en clases.

` ` 
Puede ser una buena idea si se eiminan palabras que aporten poco al resultado, sin embargo, estaría añadiendo algún tipo de sesgo humano en el algoritmo.  
De todas formas, una lista de palabras útiles a eliminar sería del estilo ["y", "o", "u", "el", "la", ...] que incluya mayormente artículos y palabras similares.
` ` 

**Pregunta 5 (0.6 puntos): Acorde al paper [A Vector Space Model for automatic indexing](https://dl.acm.org/doi/pdf/10.1145/361219.361220) un documento, $D_i$, puede ser definido formalmente como una tupla de términos, $(d_{i1}, d_{i2}, ..., d_{in})$, donde $d_{ij}$ representa el paso del j-esimo término. En clase vieron algunas formas medir los pesos de estos términos. Mencione cuales fueron y sus ventajas y desventajas.** 

` ` 
Los pesos de los términos pueden ser calculados como $w(t_i,d_j) = tf_i \cdot log_{10} (N/n_i)$  
Una ventaja de este modelo es que da importancia a palabras de pocas apariciones pero que, a la vez, le quita importancia si ha sido repetida muchas veces.  
Una desventaja es que cada vector puede tener demasiadas dimensiones nulas por o que habría que regularizar este valor, por ejemplo, sumando y/o multiplicando por una constante.
` ` 

## **2 - Preguntas prácticas 💻 (4 puntos).** ##

Esta segunda sección incluye ejercicios de programación 🤙. Leer atentamente las instrucciones entregadas a continuación para facilitar el proceso de revisión de sus trabajos.


` ` 

**Instrucciones:**



- Escribe tu código entre las lineas de comentarios **### Aquí inicia tu código ###** y **### Aquí termina tu código ###**.
- Cuando el ejercicio incluya un bloque llamado ***Test***, comprueba que el resultado de la ejecución coincida con el resultado esperado.
- Recuerde siempre mantener buenas prácticas de código.
- Está permitido sólo utilizar las librerías importadas antes del Ejercicio 1.
- **Recordar** que: *Documento = Oración. Dataset = Corpus. Vocabulario = Tokens*.
- El **orden de los resultados** pueden variar dependiendo de su máquina, pero los valores de los resultados son los mismos.



**Ejemplo:** Implemente una función **`hello_world()`** que imprima en pantalla `"Hello World"`. 

In [120]:
def hello_world():
  ### Aquí inicia tu código ###
  print("Hello World")
  ### Aquí termina tu código ###

***Test:***

In [121]:
hello_world()

Hello World


***Resultado esperado***: 
<table>
    <tr> 
        <td> Hello World </td> 
    </tr>
</table> 

``
``




Estas son las librerías permitidas. Si quieren utilizar alguna librería adicional, pueden realizar la consulta a través de Discord. 

In [122]:
import codecs
import re
import numpy as np
import pandas as pd

En caso de desarrollar la tarea desde colab, utilizar el siguiente código para cargar los archivos desde el drive:

In [123]:
try:
    from google.colab import drive

    drive.mount("/content/drive")
    path = 'path/to/marcianeke.txt'
except: 
    print('Ignorando conexión drive-colab')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


` ` 

**Ejercicio 1 - *Tokenización* (0.5 puntos).** 
` `  
` ` 

En el primer ejercicio veremos la dificultad 😨 de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo. 

El archivo adjunto al enunciado de la tarea contiene la letra de una canción del marcianeke 👽. Utilice este texto para realizar su primera tokenización y ver qué tan bien funciona su función. 

Ejecute el código a continuación para cargar el ejemplo. Recuerde realizar la modificación al directorio en caso que el archivo no se encuentre en el mismo directorio del Jupyter Notebook.

In [124]:
text = codecs.open('/content/drive/MyDrive/marcianeke.txt', 'r', 'UTF-8').read()
print(text)

Brr
Marcianeke
Vamo' a estar con Pailita
Dimelo má
Ando en busca de una criminal (ah, ah)
Esa que el gatillo le gusta jalar (rata-ta)
Que le guste flotar y fumar (brr)
Tussi, keta quiere' mezclar
Dimelo má

Ando en busca de una criminal (ah, ah)
Esa que el gatillo le gusta jalar (rata-ta)
Que le guste flotar y fumar
Tussi, keta pura quiere' mezclar
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má

Esperame que ahora entro yo
Y lo que pide yo lo traje
No visto de traje
Puro corte calle, no de maquillaje
Pronto coronamos y nos vamo' de viaje
Tanto hit que hago que lo' culo bajen
Ella se va de shopping
Sale positivo si se hace el doping
Baila twerk con un poco de popping
Los fardos en el botín

Si quieren letra llamen pa' mi booking
Generando, sigo en la mía lowkey
Cooking en el estudio con tu woman
Tanto whisky, pisco que hasta lo' vecinos toman
Si se tiran pa' aca puede que la arena coman
Ja, en el chanteo titulado sin diploma
Di-dimelo má
Di-di

Implementen una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Son libres de elegir la forma de tokenizar mientras no utilicen librerías con tokenizadores ya implementados. Pueden utilizar la librería **re** importada para trabajar símbolos.


Ejemplo de uso:

`get_tokens('Este es un ejemplo de prueba.')` 

Nos entregaría:

`['Este', 'es', 'un', 'ejemplo', 'de', 'prueba', '.']`

In [125]:
def get_tokens(text):
  ### Inicio del código ###

  lista_split = text.split()
  lista_tokens=[]
  for i in lista_split:
    if ('(' or ')' or '-' or '\'' or ',') in i:
      palabra=''
      for indice in range(0, len(i)):
        if i[indice]=='(' or i[indice]==')' or i[indice]=='-' or i[indice]=='\'' or i[indice]==',':
          lista_tokens.append(i[indice])
        else:
          palabra=palabra+i[indice]
          if i[indice+1]=='(' or i[indice+1]==')' or i[indice+1]=='-' or i[indice+1]=='\'' or i[indice+1]==',':
            lista_tokens.append(palabra)
    else:
      lista_tokens.append(i)

  return lista_tokens

  ### Fin del código ###

In [126]:
tokens = get_tokens(text)
tokens

['Brr',
 'Marcianeke',
 "Vamo'",
 'a',
 'estar',
 'con',
 'Pailita',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah)',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ratata',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 '(',
 'brr',
 ')',
 'Tussi,',
 'keta',
 "quiere'",
 'mezclar',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah)',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ratata',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 'Tussi,',
 'keta',
 'pura',
 "quiere'",
 'mezclar',
 'Di-dimelo',
 'má',
 'Di-dimelo',
 'má',
 'Di-dimelo',
 'má',
 'Di-dimelo',
 'má',
 'Di-dimelo',
 'má',
 'Di-dimelo',
 'má',
 'Di-dimelo',
 'má',
 'Esperame',
 'que',
 'ahora',
 'entro',
 'yo',
 'Y',
 'lo',
 'que',
 'pide',
 'yo',
 'lo',
 'traje',
 'No',
 'visto',
 'de',
 'traje',
 'Puro',
 'corte',
 'calle,',

**Describa cuáles fueron sus supuestos para realizar la tokenización y compare sus tokens con los entregados por la librería nltk en el bloque de código de más abajo.**

Sea realizan los siguientes supuestos:  
* Las palabras son separadas por un espacio en blanco
* Los simbolos son parte de una palabra a menos que haya sido separada por un espacio de esta

Comoconsideración, se toma en cuenta que los simbolos son partes de las palabras ya que le dan un significado a esta, por ejemplo, el apóstrofe en la palabra " vamo' " hace cambiar la pronunciaciónpor lo que no se toma en cuenta como otro token.

In [127]:
from nltk.tokenize import wordpunct_tokenize 
nltk_tokens = wordpunct_tokenize(text)
nltk_tokens

['Brr',
 'Marcianeke',
 'Vamo',
 "'",
 'a',
 'estar',
 'con',
 'Pailita',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 '(',
 'brr',
 ')',
 'Tussi',
 ',',
 'keta',
 'quiere',
 "'",
 'mezclar',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 'Tussi',
 ',',
 'keta',
 'pura',
 'quiere',
 "'",
 'mezclar',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Esperame',
 'que',
 'ahora',
 'entro',
 'yo',
 'Y',
 'lo',
 'que',

` `  
` ` 

**Ejercicio 2 - *Stopwords y Stemming* (1 punto).** 
` `  
` ` 

Considere el siguiente corpus:


In [128]:
dataset = ["I like human languages", "I like programming languages", "Spanish is my favorite language"]

Diseñe una función **`get_vocab()`** que extraiga los tokens de este corpus solamente tokenizando. Puede utilizar la función del Ejercicio 1.

In [129]:
def get_vocab(dataset):
  ### Aquí inicia tu código ###
  
  unicos = []
  for frase in dataset:
    tokens = get_tokens(frase)
    for token in tokens:
      if token not in unicos:
        unicos.append(token)
  return unicos

  ### Aquí termina tu código ###

***Test:***

In [130]:
vocab = get_vocab(dataset)
vocab

['I',
 'like',
 'human',
 'languages',
 'programming',
 'Spanish',
 'is',
 'my',
 'favorite',
 'language']

``
``

***Resultado esperado***: 
<table>
    <tr> 
        <td>['favorite',
 'Spanish',
 'language',
 'I',
 'like',
 'programming',
 'languages',
 'my',
 'human',
 'is'] </td> 
    </tr>
</table> 

``
``

Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario actualizado. 

    Explique sus reglas aquí:

El idioma que se trabaja para las stemming words y stopwords es el inglés. Así, reglas convenientes pueden ser las siguientes:  

* Si la palabra termina en "S", se elimina.
* Si la palabra termina en "IES", se reemplaza por "I"
* Si la palabra termina en "SSES", se reemplaza por "SS"

Notar que estas son reglas básicas y que se debería mejorar para hacer un preprocesado más robusto.  

Sobre las stepwords, se generará una lista de palabras que se piensa no aportan debido a la cantidad de apariciones que tienen:  
[ a, an, the, and, it, for, or, but, in, my, your, our, their ]

In [131]:
def pre_processing(vocab):
  ### Aquí inicia tu código ###
  
  stop = ['a', 'an', 'the', 'and', 'it', 'for', 'or', 'but', 'in', 'my', 'your', 'our', 'their']
  S = []
  for v in vocab:
    if v not in stop:
      if v[-1] == 's':
        stemm = v[:-1]
        if stemm not in S:
          S.append(stemm)
      elif v[-4:] == 'ies':
        stemm = v[:-2]
        if stemm not in S:
          S.append(stemm)
      elif v[-4:] == 'sses':
        stemm = v[:-4]
        if stemm not in S:
          S.append(stemm)
      else:
        if v not in S:
          S.append(v)
  
  return S

  ### Aquí termina tu código ###
    

In [132]:
vocab = pre_processing(vocab)
vocab

['I', 'like', 'human', 'language', 'programming', 'Spanish', 'i', 'favorite']


**Ejercicio 3 - *Bag of Words* 🐶🐈(0.5 puntos).** 
 


Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:

**disclaimer: El orden de los resultados pueden variar**

In [133]:
d0 = 'El perro se come la comida y después se duerme'
d1 = 'El perro se despierta y después empieza a ladrar'
d2 = 'El perro ladra y después se come la comida'
d3 = 'El gato se come la comida y después se duerme'
d4 = 'El gato se despierta y después empieza a maullar'
d5 = 'El gato maulla y después se come la comida'
dataset = [d0, d1, d2, d3, d4, d5]

El objetivo de este ejercicio es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica TF-IDF. 

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es el **Bag of Words**, método mediante el cuál se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que reciba como input un arreglo de documentos y devuelva un pandas dataframe con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representa las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el bow de un determinado documento.


Por ejemplo para el siguiente dataset: 

```
dataset = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |



In [134]:
def bag_of_words(dataset):
  ### Aquí inicia tu código ###
  
  vocab = get_vocab(dataset)
  df = pd.DataFrame(columns=vocab, index= ['d0', 'd1', 'd2', 'd3', 'd4', 'd5'])
  for col in df.columns:
    df[col].values[:] = 0

  for i in [0,1,2,3,4,5]: 
    for v in vocab:
      if v in dataset[i]:
        repeticiones = dataset[i].split().count(v)
        df[v]["d"+str(i)] = repeticiones
  return df

  ### Aquí termina tu código ###

***Test:***

In [135]:
dataset_bow = bag_of_words(dataset)
dataset_bow

El perro se come la comida  y después duerme despierta empieza  a ladrar  \
d0  1     1  2    1  1      1  1       1      1         0       0  0      0   
d1  1     1  1    0  0      0  1       1      0         1       1  1      1   
d2  1     1  1    1  1      1  1       1      0         0       0  0      0   
d3  1     0  2    1  1      1  1       1      1         0       0  0      0   
d4  1     0  1    0  0      0  1       1      0         1       1  1      0   
d5  1     0  1    1  1      1  1       1      0         0       0  0      0   

   ladra gato maullar maulla  
d0     0    0       0      0  
d1     0    0       0      0  
d2     1    0       0      0  
d3     0    1       0      0  
d4     0    1       1      0  
d5     0    1       0      1

``
``

***Resultado esperado***: 

|    | El | perro | se | come | la | comida | y | después | duerme | despierta | empieza | a | ladrar | ladra | gato | maullar | maulla |
|----|---:|------:|---:|-----:|---:|-------:|--:|--------:|-------:|----------:|--------:|--:|-------:|------:|-----:|--------:|-------:|
| d0 |  1 |     1 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    0 |       0 |      0 |
| d1 |  1 |     1 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      1 |     0 |    0 |       0 |      0 |
| d2 |  1 |     1 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     1 |    0 |       0 |      0 |
| d3 |  1 |     0 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      0 |
| d4 |  1 |     0 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      0 |     0 |    1 |       1 |      0 |
| d5 |  1 |     0 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      1 |

``
``

` `  
` ` 

**Ejercicio 4 - *Calcular TF* (0.5 puntos):** Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia ${max_i({\text{tf}_{i,j})}}$, donde
i corresponde al índice de las filas (bow) y j al de las columnas (palabras). Es decir, dividir cada bow en la cantidad de veces de la palabra que aparezca más veces en ese vector. 


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{max_i({\text{tf}_{i,j})}}$$

In [136]:
def calc_tf(dataset_bow):
    ### Aquí inicia tu código ###
    
    maximo = dataset_bow.max(axis=1)
    vocab = get_vocab(dataset_bow)
    df_norm = dataset_bow.copy()
    for i in range(0,len(dataset_bow.index)):
      for j in range(0, len(vocab)):
        df_norm[vocab[j]][i] = dataset_bow[vocab[j]][i]/maximo[i]

    return df_norm

    ### Aquí termina tu código ###

***Test:***

In [137]:
tf = calc_tf(dataset_bow)
tf

El perro   se come   la comida    y después duerme despierta empieza  \
d0  0.5   0.5  1.0  0.5  0.5    0.5  0.5     0.5    0.5       0.0     0.0   
d1  1.0   1.0  1.0  0.0  0.0    0.0  1.0     1.0    0.0       1.0     1.0   
d2  1.0   1.0  1.0  1.0  1.0    1.0  1.0     1.0    0.0       0.0     0.0   
d3  0.5   0.0  1.0  0.5  0.5    0.5  0.5     0.5    0.5       0.0     0.0   
d4  1.0   0.0  1.0  0.0  0.0    0.0  1.0     1.0    0.0       1.0     1.0   
d5  1.0   0.0  1.0  1.0  1.0    1.0  1.0     1.0    0.0       0.0     0.0   

      a ladrar ladra gato maullar maulla  
d0  0.0    0.0   0.0  0.0     0.0    0.0  
d1  1.0    1.0   0.0  0.0     0.0    0.0  
d2  0.0    0.0   1.0  0.0     0.0    0.0  
d3  0.0    0.0   0.0  0.5     0.0    0.0  
d4  1.0    0.0   0.0  1.0     1.0    0.0  
d5  0.0    0.0   0.0  1.0     0.0    1.0

``
``

***Resultado esperado***: 

|    |  El | perro |  se | come |  la | comida |   y | después | duerme | despierta | empieza |   a | ladrar | ladra | gato | maullar | maulla |
|----|----:|------:|----:|-----:|----:|-------:|----:|--------:|-------:|----------:|--------:|----:|-------:|------:|-----:|--------:|-------:|
| d0 | 0.5 |   0.5 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d1 | 1.0 |   1.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    1.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d2 | 1.0 |   1.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   1.0 |  0.0 |     0.0 |    0.0 |
| d3 | 0.5 |   0.0 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.5 |     0.0 |    0.0 |
| d4 | 1.0 |   0.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    0.0 |   0.0 |  1.0 |     1.0 |    0.0 |
| d5 | 1.0 |   0.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  1.0 |     0.0 |    1.0 |

``
``

` `  
` ` 

**Ejercicio 5 - *Calcular IDF* (0.5 punto):**


Implementar `calc_idf(dataset_bow)`. Este debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el calculo de cada idf por palabra.

Recordar que $idf_{t_i} = log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ Número de documentos que contienen la palabra $t_i$ 

In [138]:
def calc_idf(dataset_bow):
    ### Aquí inicia tu código ###
    
    dicc = {}
    vocab = get_vocab(dataset_bow)
    for j in vocab:
      N = len(dataset_bow[j])
      n_i = 0
      for i in dataset_bow[j]:
        n_i+=i
        #if n_i >= 6:
        #  n_i=6
        #  break
      dicc[j] = max(np.log10(N/n_i), 0.0)
    return dicc

    ### Aquí termina tu código ###

***Test:***

In [139]:
idf = calc_idf(dataset_bow)
idf

{'El': 0.0,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'come': 0.17609125905568124,
 'la': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'y': 0.0,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'despierta': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'a': 0.47712125471966244,
 'ladrar': 0.7781512503836436,
 'ladra': 0.7781512503836436,
 'gato': 0.3010299956639812,
 'maullar': 0.7781512503836436,
 'maulla': 0.7781512503836436}

***Resultado esperado***: 

```python
{'El': 0.0, 
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0}
```

Puede notar el bajo puntaje otorgado a las palabras que más se repiten! 😮

**Ejercicio 6 - *Calcular TF-IDF & concluir similitud de documentos.* (1 punto)**


Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [140]:
def calc_tf_idf(tf, idf):
    ### Aquí inicia tu código ###

    df = tf.copy()
    vocab = get_vocab(df)
    for i in range(0,len(df.index)):
      for j in range(0, len(vocab)):
        df[vocab[j]][i] = tf[vocab[j]][i]*idf[vocab[j]]
    return df

    ### Aquí termina tu código ### 

***Test.***

In [141]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

El     perro   se      come        la    comida    y después    duerme  \
d0  0.0  0.150515  0.0  0.088046  0.088046  0.088046  0.0     0.0  0.238561   
d1  0.0   0.30103  0.0       0.0       0.0       0.0  0.0     0.0       0.0   
d2  0.0   0.30103  0.0  0.176091  0.176091  0.176091  0.0     0.0       0.0   
d3  0.0       0.0  0.0  0.088046  0.088046  0.088046  0.0     0.0  0.238561   
d4  0.0       0.0  0.0       0.0       0.0       0.0  0.0     0.0       0.0   
d5  0.0       0.0  0.0  0.176091  0.176091  0.176091  0.0     0.0       0.0   

   despierta   empieza         a    ladrar     ladra      gato   maullar  \
d0       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
d1  0.477121  0.477121  0.477121  0.778151       0.0       0.0       0.0   
d2       0.0       0.0       0.0       0.0  0.778151       0.0       0.0   
d3       0.0       0.0       0.0       0.0       0.0  0.150515       0.0   
d4  0.477121  0.477121  0.477121       0.0       0.0   0.30103  0.778151   
d5       0.0       0.0       0.0       0.0       0.0   0.30103       0.0   

      maulla  
d0       0.0  
d1       0.0  
d2       0.0  
d3       0.0  
d4       0.0  
d5  0.778151

``
``

***Resultado esperado***: 

|    |  El |    perro |  se |     come |       la |   comida |   y | después |   duerme | despierta |  empieza |        a |   ladrar |    ladra |     gato |  maullar |   maulla |
|----|----:|---------:|----:|---------:|---------:|---------:|----:|--------:|---------:|----------:|---------:|---------:|---------:|---------:|---------:|---------:|---------:|
| d0 | 0.0 | 0.150515 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d1 | 0.0 | 0.301030 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.778151 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d2 | 0.0 | 0.301030 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.778151 | 0.000000 | 0.000000 | 0.000000 |
| d3 | 0.0 | 0.000000 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.150515 | 0.000000 | 0.000000 |
| d4 | 0.0 | 0.000000 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.000000 | 0.000000 | 0.301030 | 0.778151 | 0.000000 |
| d5 | 0.0 | 0.000000 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.301030 | 0.000000 | 0.778151 |


``
``

Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica. Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos vectores. Concluya cuáles son los dos documentos más similares.

In [142]:
def cosine_similarity(v1, v2):
  ### Aquí inicia tu código ###
  
  cs = np.dot(v1, v2)/(np.linalg.norm(v1)*np.linalg.norm(v2))
  return cs

  ### Aquí termina tu código ### 

similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity


In [143]:
print(similarity_matrix)

[[1.         0.12032418 0.3223436  0.77967014 0.         0.1632828 ]
 [0.12032418 1.         0.08686457 0.         0.4952126  0.        ]
 [0.3223436  0.08686457 1.         0.1632828  0.         0.11787732]
 [0.77967014 0.         0.1632828  1.         0.12032418 0.3223436 ]
 [0.         0.4952126  0.         0.12032418 1.         0.08686457]
 [0.1632828  0.         0.11787732 0.3223436  0.08686457 1.        ]]


Observando las filas de la matriz de similaridad (o columnas pues es simétrica), vemos que el mayor número sin contar el 1 (pues compara el texto consigo mismo) es 0.77, lo cual indica la mayor similaridad entre el texto d0 y d3.

![gato](https://live.staticflickr.com/4652/38904147065_0b6c446945_b.jpg)

**Cualquier recomendación que nos quisieran dar para una futura tarea es bienvenid@!**